stac_fastapi/pgstac/stac_fastapi/pgstac/utils.py

```diff
@@ -0,0 +1,120 @@
+"""stac-fastapi utility methods."""
+from copy import deepcopy
+from typing import Any, Dict, Optional, Set, Union
+
+from stac_fastapi.pgstac.types.base_item_cache import BaseItemCache
+from stac_fastapi.types.stac import Item
+
+
+async def hydrate(
+    item: Union[Item, Dict[str, Any]], base_item_cache: BaseItemCache
+) -> Item:
+    """Hydrate item in-place with base_item properties.
+
+    This will not perform a deep copy; values of the original item will be referenced
+    in the return item.
+    """
+    item = dict(item)
+
+    # Merge will mutate i, but create deep copies of values in the base item
+    # This will prevent the base item values from being mutated, e.g. by
+    # filtering out fields in `filter_fields`.
+    def merge(b: Dict[str, Any], i: Dict[str, Any]):
+        for key in b:
+            if key in i:
+                if isinstance(b[key], dict) and isinstance(i.get(key), dict):
+                    # Recurse on dicts to merge values
+                    merge(b[key], i[key])
+                elif b[key] == i.get(key):
+                    # Matching key/value is a no-op
+                    pass
+                elif isinstance(b[key], list) and isinstance(i.get(key), list):
+                    # Merge unequal lists
+                    i[key].extend(deepcopy(b[key]))
+            else:
+                # Keys in base item that are not in item are simply copied over
+                i[key] = deepcopy(b[key])
+
+    base_item = await base_item_cache.get(item["collection"])
+    merge(base_item, item)
+    return Item(**item)
```

Comment:
Just confirming that we need a `pystac.Item` here, rather than a regular `dict`?

I'm not familiar with stac-fastapi so maybe we already used items in some places. We'll need to be careful with performance here, to make sure no expensive operations are happening in `Item.__init__`.

Reply:
Technically, that [`Item` was a `TypedDict`](https://github.com/stac-utils/stac-fastapi/blob/master/stac_fastapi/types/stac_fastapi/types/stac.py#L49) and not a `pystac.Item` -- but yes, I had some concern if even that conversion was an unneccessary cost (though I suspect not much). Regardless, after migrating the function to pypgstac, it operates only on `Dict[str, Any]` and doesn't initialize any other types.